zigpy/ota/provider.py

```diff
@@ -535,3 +536,85 @@ async def refresh_firmware_list(self) -> None:
                     self._cache[img.key] = img
 
         self.update_expiration()
+
+
+@attr.s
+class INOVELLIImage:
+    manufacturer_id = attr.ib()
+    image_type = attr.ib()
+    version = attr.ib(default=None)
+    image_size = attr.ib(default=None)
+    url = attr.ib(default=None)
+
+    @classmethod
+    def new(cls, data):
+        mod = data["model"]
+        ver = data["version"]
+        url = data["firmware"]
+
+        res = cls(
+            manufacturer_id=Inovelli.MANUFACTURER_ID, model=mod, version=ver, url=url
+        )
+        return res
+
+    @property
+    def key(self):
+        return ImageKey(self.manufacturer_id, self.image_type)
+
+    async def fetch_image(self) -> BaseOTAImage | None:
+        async with aiohttp.ClientSession() as req:
+            LOGGER.debug("Downloading %s for %s", self.url, self.key)
+            async with req.get(self.url) as rsp:
+                data = await rsp.read()
+
+        ota_image, _ = parse_ota_image(data)
+        assert ota_image.header.key == self.key
+
+        LOGGER.debug(
+            "Finished downloading %s bytes from %s for %s ver %s",
+            self.image_size,
+            self.url,
+            self.key,
+            self.version,
+        )
+        return ota_image
+
+
+class Inovelli(Basic):
+    """Inovelli OTA Firmware provider."""
+
+    UPDATE_URL = "https://files.inovelli.com/firmware/firmware.json"
+    MANUFACTURER_ID = 4655
+    HEADERS = {"accept": "application/json"}
+
+    async def initialize_provider(self, ota_config: dict) -> None:
+        self.info("OTA provider enabled")
+        self.config = ota_config
+        await self.refresh_firmware_list()
+        self.enable()
+
+    async def refresh_firmware_list(self) -> None:
+        if self._locks[LOCK_REFRESH].locked():
+            return
+
+        async with self._locks[LOCK_REFRESH]:
+            async with aiohttp.ClientSession(headers=self.HEADERS) as req:
+                async with req.get(self.UPDATE_URL) as rsp:
+                    if not (200 <= rsp.status <= 299):
+                        self.warning(
+                            "Couldn't download '%s': %s/%s",
+                            rsp.url,
+                            rsp.status,
+                            rsp.reason,
+                        )
+                        return
+                    fw_lst = await rsp.json()
+        self.debug("Finished downloading firmware update list")
+        self._cache.clear()
+        for fw in fw_lst:
```

Comment:
The response JSON looks like a dictionary, not a list, so iterating over it will yield only the string keys:

```
{
    device.model: [list of firmware objects]
}
```

Since the image objects themselves don't contain enough information to identify the target device itself from just its `query_next_image` request, you will have to modify `INOVELLIImage.new` to contain a second parameter for the device model name (like Salus). This should work:

```Python
for model, firmwares in firmwares.items():
    # Pick the most recent firmware
    firmware = max(firmwares, key=lambda obj: obj["version"])
    img = INOVELLIImage.new(data=firmware, model=model)
```

If you are able to add new keys to the JSON, it would be easiest to match the data that your device sends when requesting a new image:

```Python
query_next_image(field_control=0, manufacturer_code=4655, image_type=0, current_file_version=9, hardware_version=None)
```

Ideally, every image in the JSON would be uniquely identified by its `image_type`, `manufacturer_code`, and `file_version`. The manufacturer code seems to be static and the `file_version` decimal is encoded as a string so it would be best to add a `image_type` key.

If this isn't possible, you will have to basically implement the `SalusImage` above, which uses the device model name instead of its `image_type`.

Reply:
I believe I've updated this correctly, but I'm getting this error with the new parameter -
```"WARNING (MainThread) [zigpy.util] Error calling listener: local variable 'firmwares' referenced before assignment"```

This is also what I'm seeing in my debug logs when issuing the cluster command after making these changes - 
```2022-04-26 20:53:56 DEBUG (MainThread) [homeassistant.components.zha.api] Issued command for: cluster_id: [25] cluster_type: [out] endpoint_id: [1] command: [0] command_type: [client] args: [0, 100] manufacturer: [None] response: TXStatus.SUCCESS

2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Received ZCL frame: b'\x11#\x01\x00/\x12\x00\x00\t\x00\x00\x00'
2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.CLUSTER_COMMAND: 1>, is_manufacturer_specific=0, is_reply=0, disable_default_response=1, reserved=0, *is_cluster=True, *is_general=False), tsn=35, command_id=1, *is_reply=False)
2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Decoded ZCL frame: Ota:query_next_image(field_control=0, manufacturer_code=4655, image_type=0, current_file_version=9, hardware_version=None)
2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Received command 0x01 (TSN 35): query_next_image(field_control=0, manufacturer_code=4655, image_type=0, current_file_version=9, hardware_version=None)
2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] OTA query_next_image handler for 'Inovelli VZM31-SN': field_control=0, manufacture_id=4655, image_type=0, current_file_version=9, hardware_version=None, model=VZM31-SN
2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] No OTA image is available
2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Sending reply header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.CLUSTER_COMMAND: 1>, is_manufacturer_specific=False, is_reply=1, disable_default_response=1, reserved=0, *is_cluster=True, *is_general=False), tsn=35, command_id=2, *is_reply=True)
2022-04-26 20:53:56 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Sending reply: query_next_image_response(status=<Status.NO_IMAGE_AVAILABLE: 152>, manufacturer_code=None, image_type=None, file_version=None, image_size=None)
```
Something that stands out to me is that the query_next_image handler is listing it as "manufacture_id=4655" as opposed to manufacturer_id=4655, would that be causing any issues?

Reply:
My bad, it should be `fw_lst.items()` instead of `firmwares.items()`.

The parameter name discrepancy is just a typo in the logging string, it has no impact on anything else. If the OTA provider fails to refresh the firmware list due to the above error, no OTA image will exist for the device to use so that's expected behavior.

Reply:
Thanks! Looks like now it's throwing - 
```
WARNING (MainThread) [zigpy.util] Error calling listener: __init__() got an unexpected keyword argument 'model'
```
Is there any way to tell which reference to model is unexpected? My assumption is that it would have to come after the previous fw_lst.items() error, which means it's part of
``            img = INOVELLIImage.new(data=firmware, model=model) ``

Reply:
Also is it a correct understanding that you're saying that while this will throw the warning log, it's just the logging itself, and I should be seeing this come through successfully?

Reply:
You'll have to add a `model` attribute to the class if you're referencing it in the initializer. The following changes to your patch get zigpy to start up for me:

```diff
diff --git a/zigpy/ota/provider.py b/zigpy/ota/provider.py
index 009dd2c..cd1ecb9 100644
--- a/zigpy/ota/provider.py
+++ b/zigpy/ota/provider.py
@@ -541,6 +541,7 @@ class FileStore(Basic):
 @attr.s
 class INOVELLIImage:
     manufacturer_id = attr.ib()
+    model = attr.ib()
     version = attr.ib(default=None)
     url = attr.ib(default=None)

@@ -611,6 +612,8 @@ class Inovelli(Basic):
             # Pick the most recent firmware
             firmware = max(firmwares, key=lambda obj: obj["version"])
             img = INOVELLIImage.new(data=firmware, model=model)
+            self._cache[img.key] = img
+
         self.update_expiration()

     async def filter_get_image(self, key: ImageKey) -> bool:
```

Reply:
Thank you so much! It clearly recognizes that it needs firmware v10, and gets the URL for it, but then it fails with only a "Error calling listener:" - 
```
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.CLUSTER_COMMAND: 1>, is_manufacturer_specific=0, is_reply=0, disable_default_response=1, reserved=0, *is_cluster=True, *is_general=False), tsn=111, command_id=1, *is_reply=False)
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Decoded ZCL frame: Ota:query_next_image(field_control=0, manufacturer_code=4655, image_type=0, current_file_version=9, hardware_version=None)
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Received command 0x01 (TSN 111): query_next_image(field_control=0, manufacturer_code=4655, image_type=0, current_file_version=9, hardware_version=None)
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] OTA query_next_image handler for 'Inovelli VZM31-SN': field_control=0, manufacture_id=4655, image_type=0, current_file_version=9, hardware_version=None, model=VZM31-SN
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.ota.provider] Downloading https://files.inovelli.com/firmware/VZM31-SN/Beta/00000010/00000010.ota for ImageKey(manufacturer_id=4655, image_type='VZM31-SN')
2022-04-27 08:11:32 WARNING (MainThread) [zigpy.util] Error calling listener:
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] No OTA image is available
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Sending reply header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.CLUSTER_COMMAND: 1>, is_manufacturer_specific=False, is_reply=1, disable_default_response=1, reserved=0, *is_cluster=True, *is_general=False), tsn=111, command_id=2, *is_reply=True)
2022-04-27 08:11:32 DEBUG (MainThread) [zigpy.zcl] [0xAB52:1:0x0019] Sending reply: query_next_image_response(status=<Status.NO_IMAGE_AVAILABLE: 152>, manufacturer_code=None, image_type=None, file_version=None, image_size=None)
```
Does zigpy save the OTA files somewhere when it downloads them so I could check if it made any progress with the download? Wireshark showed a decent amount of back and forth between the server and the firmware repo, so not sure if I successfully downloaded the full ota and then something failed after that?

Reply:
I've submitted a PR to your zigpy fork that should fix this problem: https://github.com/codyhackw/zigpy/pull/1.

Thanks for spotting the `Error calling listener:` logging bug, it definitely isn't making development easier by discarding the traceback.

Reply:
That did it, the OTA update is working as I write this. Your help has been fantastic and I really appreciate it! I know fellow end users will as well!